File: csi_camera_ros2/csi_camera_ros2/csi_camera_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy
import cv2
from sensor_msgs.msg import Image, CompressedImage
from cv_bridge import CvBridge

# Revert pipeline EXACTLY to the working example script's structure and formatting
def gstreamer_pipeline(
    sensor_id=0,
    capture_width=640, # Keep VGA default
    capture_height=480, # Keep VGA default
    display_width=640, # Match capture for ROS node
    display_height=480, # Match capture for ROS node
    framerate=20, # Keep 20fps default
    flip_method=0, # Keep flip_method=0
):
    return (
        "nvarguscamerasrc sensor-id=%d ! "
        "video/x-raw(memory:NVMM), width=(int)%d, height=(int)%d, framerate=(fraction)%d/1 ! "
        "nvvidconv flip-method=%d ! "
        "video/x-raw, width=(int)%d, height=(int)%d, format=(string)BGRx ! "
        "videoconvert ! "
        "video/x-raw, format=(string)BGR ! appsink" # Removed drop=true
        % (
            sensor_id,
            capture_width,
            capture_height,
            framerate,
            flip_method,
            display_width, # Use display_width here
            display_height, # Use display_height here
        )
    )

class CSICameraNode(Node):
    def __init__(self):
        super().__init__('csi_camera_node')

        # Declare parameters
        self.declare_parameter('sensor_id', 0)
        self.declare_parameter('sensor_mode', 0) # Using mode 0
        self.declare_parameter('capture_width', 640)  # Default VGA
        self.declare_parameter('capture_height', 480) # Default VGA
        # Add display width/height params used in the pipeline function now
        self.declare_parameter('display_width', 640)
        self.declare_parameter('display_height', 480)
        self.declare_parameter('framerate', 20)       # Default 20fps
        self.declare_parameter('flip_method', 0)      # Default flip_method
        self.declare_parameter('publish_rate', 20.0)  # Match framerate
        self.declare_parameter('publish_compressed', True) # Publish compressed by default

        # Get parameters
        self.sensor_id = self.get_parameter('sensor_id').get_parameter_value().integer_value
        self.sensor_mode = self.get_parameter('sensor_mode').get_parameter_value().integer_value
        self.capture_width = self.get_parameter('capture_width').get_parameter_value().integer_value
        self.capture_height = self.get_parameter('capture_height').get_parameter_value().integer_value
        # Get display width/height params
        self.display_width = self.get_parameter('display_width').get_parameter_value().integer_value
        self.display_height = self.get_parameter('display_height').get_parameter_value().integer_value
        self.framerate = self.get_parameter('framerate').get_parameter_value().integer_value
        self.flip_method = self.get_parameter('flip_method').get_parameter_value().integer_value
        publish_rate = self.get_parameter('publish_rate').get_parameter_value().double_value
        self.publish_compressed = self.get_parameter('publish_compressed').get_parameter_value().bool_value

        # Define a QoS profile: Best Effort, Depth 1
        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=1
        )

        topic_name_raw = f'/csi_camera_{self.sensor_id}/image_raw'
        self.image_pub_raw = self.create_publisher(Image, topic_name_raw, qos_profile) # Use QoS profile
        self.image_pub_compressed = None
        log_message = f"Started camera node for sensor ID {self.sensor_id}. Publishing Raw: {topic_name_raw}"
        if self.publish_compressed:
            topic_name_compressed = f'{topic_name_raw}/compressed'
            self.image_pub_compressed = self.create_publisher(CompressedImage, topic_name_compressed, qos_profile) # Use QoS profile
            log_message += f", Compressed: {topic_name_compressed}"
        self.bridge = CvBridge()

        self.pipeline = gstreamer_pipeline(
            # Pass all params to the pipeline function now
            sensor_id=self.sensor_id,
            sensor_mode=self.sensor_mode,
            capture_width=self.capture_width,
            capture_height=self.capture_height,
            display_width=self.display_width,
            display_height=self.display_height,
            framerate=self.framerate,
            flip_method=self.flip_method
        )
        self.get_logger().info(f"Using GStreamer pipeline: {self.pipeline}") # Log the exact pipeline being used

        self.cap = cv2.VideoCapture(self.pipeline, cv2.CAP_GSTREAMER)

        if not self.cap.isOpened():
            self.get_logger().error(f"Error: Unable to open camera with sensor ID {self.sensor_id}")
            rclpy.shutdown()
            return

        self.timer = self.create_timer(1.0 / publish_rate, self.timer_callback)
        self.get_logger().info(log_message) # Use the constructed log message

    def timer_callback(self):
        ret, frame = self.cap.read()
        if not ret:
            self.get_logger().error(f"Error: Unable to read frame from camera with sensor ID {self.sensor_id}")
            # Consider adding logic to attempt reconnection or shutdown
            return

        # Prepare timestamp and frame_id
        now = self.get_clock().now().to_msg()
        frame_id = f"camera_{self.sensor_id}_frame"

        # Publish raw image
        image_msg_raw = self.bridge.cv2_to_imgmsg(frame, "bgr8")
        image_msg_raw.header.stamp = now
        image_msg_raw.header.frame_id = frame_id
        self.image_pub_raw.publish(image_msg_raw)

        # Publish compressed image only if enabled
        if self.publish_compressed and self.image_pub_compressed:
            compressed_msg = CompressedImage()
            compressed_msg.header.stamp = now
            compressed_msg.header.frame_id = frame_id
            compressed_msg.format = "jpeg"
            # Use imencode parameters for potential speedup (lower quality)
            encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), 90] # Quality 90 (0-100)
            compressed_msg.data = cv2.imencode('.jpg', frame, encode_param)[1].tobytes()
            self.image_pub_compressed.publish(compressed_msg)

    def destroy_node(self):
        self.get_logger().info("Shutting down camera node.")
        if self.cap.isOpened():
            self.cap.release()
        super().destroy_node()

def main(args=None):
    rclpy.init(args=args)
    csi_camera_node = CSICameraNode()
    try:
        rclpy.spin(csi_camera_node)
    except KeyboardInterrupt:
        pass
    finally:
        # Destroy the node explicitly
        # (optional - otherwise it will be done automatically
        # when the garbage collector destroys the node object)
        csi_camera_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: csi_camera_cpp/src/csi_camera_node.cpp
#include "csi_camera_cpp/csi_camera_node.hpp" // Fix the include

#include <rclcpp_components/register_node_macro.hpp> // For composable nodes
#include <chrono>
#include <sstream> // For std::stringstream

using namespace std::chrono_literals;

namespace csi_camera_cpp
{

// GStreamer pipeline function (similar to the C++ example)
std::string CSICameraNode::gstreamer_pipeline_string()
{
    std::stringstream ss;
    ss << "nvarguscamerasrc sensor-id=" << sensor_id_ << " ! "
       << "video/x-raw(memory:NVMM), width=(int)" << capture_width_ << ", height=(int)" << capture_height_ << ", framerate=(fraction)" << framerate_ << "/1 ! "
       << "nvvidconv flip-method=" << flip_method_ << " ! "
       // Remove explicit width/height here, let videoconvert handle final format
       << "video/x-raw, format=(string)BGRx ! "
       << "videoconvert ! "
       << "video/x-raw, format=(string)BGR ! appsink max-buffers=1 sync=false"; // Keep low-latency appsink settings
    return ss.str();
}

CSICameraNode::CSICameraNode(const rclcpp::NodeOptions & options)
: Node("csi_camera_node", options) // Pass options for composable nodes
{
    RCLCPP_INFO(this->get_logger(), "Initializing CSICameraNode...");
    declare_parameters();
    load_parameters();
    setup_camera();
    setup_publisher();
    fill_camera_info(); // Initialize camera info
    setup_timer();
    RCLCPP_INFO(this->get_logger(), "CSICameraNode initialized successfully.");
}

CSICameraNode::~CSICameraNode()
{
    RCLCPP_INFO(this->get_logger(), "Shutting down CSICameraNode...");
    if (timer_) {
        timer_->cancel();
    }
    if (cap_.isOpened()) {
        cap_.release();
        RCLCPP_INFO(this->get_logger(), "Camera capture released.");
    }
    RCLCPP_INFO(this->get_logger(), "CSICameraNode shutdown complete.");
}

void CSICameraNode::declare_parameters()
{
    this->declare_parameter<int>("sensor_id", 0);
    this->declare_parameter<int>("capture_width", 1920);
    this->declare_parameter<int>("capture_height", 1080);
    this->declare_parameter<int>("display_width", 960);
    this->declare_parameter<int>("display_height", 540);
    this->declare_parameter<int>("framerate", 30);
    this->declare_parameter<int>("flip_method", 0);
    this->declare_parameter<double>("publish_rate", 30.0);
    this->declare_parameter<std::string>("frame_id", "camera_frame");
    // Add parameter for camera info URL if needed later
    // this->declare_parameter<std::string>("camera_info_url", "");
}

void CSICameraNode::load_parameters()
{
    this->get_parameter("sensor_id", sensor_id_);
    this->get_parameter("capture_width", capture_width_);
    this->get_parameter("capture_height", capture_height_);
    this->get_parameter("display_width", display_width_); // Corrected ->
    this->get_parameter("display_height", display_height_); // Corrected ->
    this->get_parameter("framerate", framerate_); // Corrected ->
    this->get_parameter("flip_method", flip_method_); // Corrected ->
    this->get_parameter("publish_rate", publish_rate_); // Corrected ->
    this->get_parameter("frame_id", frame_id_); // Corrected ->
}

void CSICameraNode::setup_camera()
{
    std::string pipeline = gstreamer_pipeline_string();
    RCLCPP_INFO(this->get_logger(), "Using GStreamer pipeline: %s", pipeline.c_str());

    RCLCPP_INFO(this->get_logger(), "Attempting to open camera...");
    cap_.open(pipeline, cv::CAP_GSTREAMER);

    // Add a small delay and check if opened
    rclcpp::sleep_for(2s); // Use rclcpp sleep

    if (!cap_.isOpened()) {
        // Throw an exception instead of shutting down directly
        std::string err_msg = "Error: Unable to open camera with GStreamer pipeline. Check pipeline string and camera connection.";
        RCLCPP_ERROR(this->get_logger(), err_msg.c_str());
        throw std::runtime_error(err_msg);
    } else {
        RCLCPP_INFO(this->get_logger(), "Successfully opened camera using GStreamer pipeline.");
    }
}

void CSICameraNode::setup_publisher()
{
    // Use image_transport to create publisher
    camera_pub_ = image_transport::create_camera_publisher(this, "image_raw");
    RCLCPP_INFO(this->get_logger(), "Publishing to %s", camera_pub_.getTopic().c_str());
    RCLCPP_INFO(this->get_logger(), "Publishing CameraInfo to %s", camera_pub_.getInfoTopic().c_str());
}

void CSICameraNode::setup_timer()
{
    if (publish_rate_ <= 0) {
        RCLCPP_WARN(this->get_logger(), "Publish rate is <= 0, timer will not be created.");
        return;
    }
    auto timer_period = std::chrono::duration<double>(1.0 / publish_rate_);
    timer_ = this->create_wall_timer(timer_period, std::bind(&CSICameraNode::timer_callback, this));
    RCLCPP_INFO(this->get_logger(), "Timer created with period: %.4f s (%.1f Hz)", timer_period.count(), publish_rate_);
}

void CSICameraNode::fill_camera_info()
{
    // Create a default CameraInfo message.
    // Ideally, load this from a calibration file using camera_info_manager
    camera_info_msg_.header.frame_id = frame_id_;
    camera_info_msg_.height = display_height_; // Use display height for info
    camera_info_msg_.width = display_width_;   // Use display width for info
    // Set default distortion model and parameters (pinhole, all zeros)
    camera_info_msg_.distortion_model = "plumb_bob";
    camera_info_msg_.d.resize(5, 0.0);
    // Set default camera matrix (K) - identity for now
    camera_info_msg_.k.fill(0.0);
    camera_info_msg_.k[0] = display_width_;  // fx approx width
    camera_info_msg_.k[4] = display_height_; // fy approx height
    camera_info_msg_.k[2] = display_width_ / 2.0; // cx
    camera_info_msg_.k[5] = display_height_ / 2.0; // cy
    camera_info_msg_.k[8] = 1.0;
    // Set default rectification matrix (R) - identity
    camera_info_msg_.r.fill(0.0);
    camera_info_msg_.r[0] = camera_info_msg_.r[4] = camera_info_msg_.r[8] = 1.0;
    // Set default projection matrix (P) - K [I|0] for now
    camera_info_msg_.p.fill(0.0);
    camera_info_msg_.p[0] = camera_info_msg_.k[0]; // fx
    camera_info_msg_.p[5] = camera_info_msg_.k[4]; // fy
    camera_info_msg_.p[2] = camera_info_msg_.k[2]; // cx
    camera_info_msg_.p[6] = camera_info_msg_.k[5]; // cy
    camera_info_msg_.p[10] = 1.0;

    RCLCPP_INFO(this->get_logger(), "Generated default CameraInfo.");
}


void CSICameraNode::timer_callback()
{
    if (!cap_.isOpened()) {
        RCLCPP_WARN(this->get_logger(), "Timer callback called but camera is not open.");
        return;
    }

    if (!cap_.read(frame_)) {
        RCLCPP_WARN(this->get_logger(), "cap.read() failed.");
        return;
    }

    if (frame_.empty()) {
         RCLCPP_WARN(this->get_logger(), "Frame is empty.");
         return;
    }

    // Convert OpenCV image to ROS Image message
    auto now = this->get_clock()->now();
    std_msgs::msg::Header header;
    header.stamp = now;
    header.frame_id = frame_id_;

    // Use cv_bridge to convert cv::Mat to sensor_msgs::msg::Image
    sensor_msgs::msg::Image::SharedPtr image_msg = cv_bridge::CvImage(header, "bgr8", frame_).toImageMsg();

    // Update timestamp in CameraInfo message
    camera_info_msg_.header.stamp = now;

    // Publish image and camera info
    camera_pub_.publish(image_msg, std::make_shared<sensor_msgs::msg::CameraInfo>(camera_info_msg_));

    // Optional: Log periodically
    // static int frame_count = 0;
    // if (frame_count++ % 60 == 0) {
    //     RCLCPP_INFO(this->get_logger(), "Published frame %d", frame_count);
    // }
}

} // namespace csi_camera_cpp

// Register the component with class_loader
RCLCPP_COMPONENTS_REGISTER_NODE(csi_camera_cpp::CSICameraNode)
